### ISSUE TITLE
Initializer should use at most 2^(n+1)-2n cnots

### ISSUE THREAD

**Issue by adjs** [Qiskit/qiskit-terra#4182]
### What is the expected enhancement?

Initializer should use at most 2^(n+1)-2n cnots (arXiv:quant-ph/0406176). 

Cost to initialize a 4 qubits state uses  
Number of cnots (optimization_level=0):  44
Number of cnots (optimization_level=3):  26
max number of cnots:  24

```python
num_qubits = 4
_optimization_level = 3

vector = -1 * np.random.rand(2 ** num_qubits)
vector[0] = 0
vector = vector / np.linalg.norm(vector)

qr = QuantumRegister(num_qubits, 'qr')
circuit = QuantumCircuit(qr)
circuit.initialize(vector, qr)

b = transpile(circuit, basis_gates=['u1', 'u2', 'u3', 'cx'], optimization_level=_optimization_level)
print('Number of cnots: ', b.count_ops()['cx'])
print('max number of cnots: ', 2 ** (num_qubits + 1) - 2 * num_qubits)
```

**Comment by Cryoris** [comment]
There's another implementation for arbitrary `m x n` isometries which can also be used for state preparation, which should be more efficient. In your example you can test it by using
```python
circuit.isometry(vector, None, qr)
```
When I'm running this I'm getting 11 CX gates, for both, optimization level 0 as well as 3. 🙂 We probably should make `initialize` use the isometry per default, if there's no other advantage to the `initialize` method.

**Comment by adjs** [comment]
@Cryoris, thanks for the reference. I'll take a look on arXiv:1501.06911 .

I tested using the isometry method and got a different result.

1. I added a seed in the first line `np.random.seed(1)`
2. changed the vector.

```python
vector = -2 * np.random.rand(2 ** num_qubits) + 1 + -2 * np.random.rand(2 ** num_qubits)*1j + 1j
```

With this configuration initialize (in PR #4183) outputs the desired vector with 22 cnots
`[ 0.        +0.j         -0.12734211-0.03392127j  0.28892154+0.20784746j
  0.11424689+0.17448988j  0.20416636-0.17382293j  0.23561822-0.27064362j
  0.18133384+0.10783621j  0.08926208-0.11115772j  0.05966585-0.21754363j
 -0.02243511-0.2280729j   0.04670357+0.23983419j -0.10705229+0.26641494j
  0.17081929+0.19082986j -0.21854253-0.21855702j  0.27315829+0.2321456j
 -0.09852601+0.04559784j]`

and isometry outputs a different vector with 11 cnots
[ 0.        +0.j          0.12463131+0.04282176j -0.27353611-0.22771522j
 -0.10165046-0.18211604j -0.21592214+0.15898409j -0.2541271 +0.25334429j
 -0.17327317-0.12036204j -0.0968827 +0.10458251j -0.07486663+0.21279152j
  0.00628679+0.22908745j -0.02966491-0.24253176j  0.12558324-0.25819753j
 -0.15692895-0.20240695j  0.20257686+0.23343227j -0.25609771-0.25084057j
  0.10149775-0.03853238j]
